core/alsp_src/generic/bsio.c

```diff
@@ -334,7 +338,46 @@ long standard_console_error(char *buf, long n)
 
 long standard_console_read(char *buf, long n)
 {
-    return read(STDIN_FILENO, buf, n);
+    char *line;
+    long count;
+
+        /* Load the history file once (i.e., at startup) 
+    	   do_load_prev_history is set == 1 by default (set == 0 during shell startup)
+	   Setting it to 0 ensures the
+	   history will only be loaded once during an alspro shell session.
+	*/
+    if (do_load_prev_history == 1){
+        linenoiseHistoryLoad(history_file); 
+        do_load_prev_history = 0;
+    }
+
+    if (console_prompt == NULL){
+	console_prompt = "?- ";
+    }
+
+    line = linenoise(console_prompt);
+    if (line == NULL){
+	return 0;
+    }
+    count = (int)strlen(line);
+		/* count+1 because '\n' may be added here: */
+    if (count+1 <= n )
+    {
+	linenoiseHistoryAdd(line);  // Add to the history. 
+    	linenoiseHistorySave(history_file);  // Save the history on disk. 
+	memcpy(buf, line, count);
+	buf[count] = '\n'; count++;
+    } 
+    else {    /* Incoming line overruns buf: */
+	errno = ENOBUFS;
+	return -1;
```

Comment:
Pedantic, but maybe set `count = -1` rather then `return -1` so that `line` gets freed before the function return.